### PR TITLE
Fix things

### PR DESCRIPTION
**Pull request by saagarjha**
Not ptrace-specific

### CODE REVIEW

File: kernel/task.c [hunk 1]
```diff
@@ -94,6 +94,11 @@ void (*task_run_hook)(void) = NULL;
 
 static void *task_run(void *task) {
     current = task;
+#if __APPLE__
```
Comment (tbodt) [inline kernel/task.c:97]: Could you move the ifdefs into a function, maybe in sync.h, so they don't have to be duplicated for both fork and exec?

File: util/sync.h [hunk 2]
```diff
@@ -40,12 +40,13 @@ static inline void __lock(lock_t *lock, __attribute__((unused)) const char *file
...
 static inline void unlock(lock_t *lock) {
-    pthread_mutex_unlock(&lock->m);
 #if LOCK_DEBUG
     lock->debug = (struct lock_debug) {};
 #endif
+    lock->owner = NULL;
+    pthread_mutex_unlock(&lock->m);
 }
-#define unlock(lock) pthread_mutex_unlock(&(lock)->m)
+#define unlock(lock) unlock(lock)
```
Comment (tbodt) [inline util/sync.h:49]: Don't think there's a reason to have this as a macro anymore

File: kernel/exit.c [hunk 3]
```diff
@@ -220,8 +220,10 @@ static bool reap_if_zombie(struct task *task, struct siginfo_ *info_out, struct
 }
 
 static bool notify_if_stopped(struct task *task, struct siginfo_ *info_out) {
-    // FIXME the check of task->group->stopped needs locking I think
-    if (!task->group->stopped || task->group->group_exit_code == 0)
+    lock(&current->group->lock);
```
Comment (tbodt) [inline kernel/exit.c:223]: Should be task->group->lock

File: app/AppGroup.m [hunk 4]
```diff
@@ -101,11 +101,11 @@
                                                                       error:nil];
 }
 
-NSArray<NSString *> *CurrentAppGroups() {
+NSArray<NSString *> *CurrentAppGroups(void) {
     return AppEntitlements()[@"com.apple.security.application-groups"];
 }
 
-NSURL *ContainerURL() {
+NSURL *ContainerURL(void) {
```
Comment (tbodt) [inline app/AppGroup.m:108]: Pretty sure (void) isn't needed for function definitions, only prototypes

Comment (saagarjha) [reply app/AppGroup.m:108]: It is not, although trying to call it incorrectly invokes undefined behavior

Comment (tbodt) [reply app/AppGroup.m:108]: The idea is you need it on the prototype so you get an error when you call it wrong, but you don't need it on the definition

Comment (saagarjha) [reply app/AppGroup.m:108]: Apparently defining it as such creates a 0-argument prototype as well, and then trying to call a function that has a prototype defined with the wrong number of arguments is undefined.

Comment (tbodt) [reply app/AppGroup.m]: Weird but ok

File: app/AppGroup.m [hunk 5]
```diff
@@ -41,14 +41,14 @@
         return nil;
     if (dl_info.dli_fbase == NULL)
         return nil;
-    uint8_t *base = dl_info.dli_fbase;
+    char *base = dl_info.dli_fbase;
```
Comment (tbodt) [inline app/AppGroup.m:44]: What does this fix?

Comment (saagarjha) [reply app/AppGroup.m:44]: `char *` and `void *` are the legal aliasing types

Comment (tbodt) [reply app/AppGroup.m:44]: Guilty as charged

Comment (saagarjha) [reply app/AppGroup.m:44]: I should note that *any* character type is a legal aliasing type, so you are also welcome to use `signed char *` and `unsigned char *` if you so wish.

**Pull request merged**